### PR TITLE
custom autograd op redispatching fix

### PR DESCRIPTION
**Pull request by bdhirsh**
This is enough to fix the XLA errors that I'm seeing in https://github.com/pytorch/pytorch/pull/106406.

There were two problems:
(1) We weren't hitting the `PythonTLSSnapshot` key when redispatching, causing this error: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/core/PythonFallbackKernel.cpp#L48

(2) The custom XLA ops weren't going through the python dispatcher, which requires the `PythonDispatcher` dispatch key. This didn't actually break anything, but made debugging harder (and a nice invariant we have anyway is that all ops go through the python dispatcher when running AOTAutograd).

I think that at some point we should revisit this interaction, and see if we can e.g. remove these extra hardcoded dispatch keys and use `Dispatcher.call()` instead`. I'd just really like to be able to land this first though so that I can finally land python functionalization (see the linked PR).

### CODE REVIEW

File: torch_xla/csrc/aten_autograd_ops.cpp [hunk 1]
```diff
@@ -141,6 +151,15 @@ torch::autograd::variable_list MaxPool2dAutogradFunction::backward(
...
   auto ks = self_keyset & mask;
+  // If python dispatcher is enabled, we need to hit it.
+  // This is a bit hacky, we should probably come up with
+  // a better way to do this (maybe don't redispatch?)
+  if (c10::impl::PythonDispatcherTLS::get_state()) {
+    ks = ks.add(c10::DispatchKey::PythonDispatcher);
+  }
+  if (ks.has(c10::DispatchKey::Python)) {
+    ks = ks.add(c10::DispatchKey::PythonTLSSnapshot);
+  }
```
Comment (JackCaoG) [inline torch_xla/csrc/aten_autograd_ops.cpp:162]: do we need to do this for all functions in this file? including `MaxPool3dAutogradFunction ` and `einsum`

**Review by JackCaoG — APPROVED** [review]
Thanks!

File: torch_xla/csrc/aten_autograd_ops.cpp (hunk 1 above)
Comment (bdhirsh) [reply torch_xla/csrc/aten_autograd_ops.cpp:162]: ah yep, I think we will. It looks like einsum and maxpool3d don't have any of the existing logic either (computing a dispatch keyset and manually reinvoking the dispatcher), so I'm guessing it's just the case that nobody has tried to used those ops with torch.compile yet.

Comment (JackCaoG) [reply torch_xla/csrc/aten_autograd_ops.cpp:162]: lol I just added a test for einsum today with `torch.compile`, should I be worried?

Comment (bdhirsh) [reply torch_xla/csrc/aten_autograd_ops.cpp:162]: Yeah…. I would naively expect it to break unless I’m missing something. I could be wrong, lmk if you hit something :)

Comment (JackCaoG) [reply torch_xla/csrc/aten_autograd_ops.cpp:162]: what's the right way for me to test it? build your branch locally?

**Pull request merged**